cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -165,68 +172,99 @@ func (gc *GceCache) updateMigBasename(ref GceRef, basename string) {
+			// It is fine that mig is not found in gc.migs map even if we have entry in gc.instanceRefToMigRef
+			// The mig could have been removed from gc.migs if we (for example) disabled autoscaling for it or the node pool
+			// was just deleted. Yet we are not removing entries from gc.instanceRefToMigRef in such case due to performance reasons.
+			return nil, nil
+		}
+		return mig.Config, nil
 	}
 
 	for _, mig := range gc.GetMigs() {
-		if mig.Config.GceRef().Project == instance.Project &&
-			mig.Config.GceRef().Zone == instance.Zone &&
-			strings.HasPrefix(instance.Name, mig.Basename) {
-			if err := gc.regenerateCache(); err != nil {
-				return nil, fmt.Errorf("error while looking for MIG for instance %+v, error: %v", *instance, err)
+		if mig.Config.GceRef().Project == instanceRef.Project &&
+			mig.Config.GceRef().Zone == instanceRef.Zone &&
+			strings.HasPrefix(instanceRef.Name, mig.Basename) {
+			if err := gc.regenerateInstanceCacheForMigNoLock(mig.Config.GceRef()); err != nil {
+				return nil, fmt.Errorf("error while looking for MIG for instance %+v, error: %v", instanceRef, err)
 			}
-			if mig, found := gc.instancesCache[*instance]; found {
-				return mig, nil
+
+			migRef, found := gc.instanceRefToMigRef[instanceRef]
+			if !found {
+				return nil, fmt.Errorf("instance %+v belongs to unknown mig %+v", instanceRef, migRef)
 			}
-			return nil, fmt.Errorf("instance %+v does not belong to any configured MIG", *instance)
+			mig, found := gc.getMig(migRef)
+			if !found {
+				return nil, fmt.Errorf("instance %+v does not belong to any configured MIG", instanceRef)
+			}
+			return mig.Config, nil
 		}
 	}
 	// Instance doesn't belong to any configured mig.
 	return nil, nil
 }
 
-// RegenerateInstancesCache triggers instances cache regeneration under lock.
-func (gc *GceCache) RegenerateInstancesCache() error {
+func (gc *GceCache) getMig(migRef GceRef) (MigInformation, bool) {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+	mig, found := gc.migs[migRef]
+	return *mig, found
+}
+
+// RegenerateInstanceCacheForMig triggers instances cache regeneration for single MIG under lock.
+func (gc *GceCache) RegenerateInstanceCacheForMig(migRef GceRef) error {
 	gc.cacheMutex.Lock()
 	defer gc.cacheMutex.Unlock()
-
-	return gc.regenerateCache()
+	return gc.regenerateInstanceCacheForMigNoLock(migRef)
 }
 
-// internal method - should only be called after locking on cacheMutex.
-func (gc *GceCache) regenerateCache() error {
-	newInstancesCache := make(map[GceRef]Mig)
+func (gc *GceCache) regenerateInstanceCacheForMigNoLock(migRef GceRef) error {
+	klog.V(4).Infof("Regenerating MIG information for %s", migRef.String())
+
+	// Here we are on purpose not deleting old entries pointing to migRef from gc.instanceRefToMig.
+	// They are no harm as same GceRef should not be used for new instance and instance should not move from one Mig to
+	// the other.
+	// And garbage-collection will be enforces as whole cache is invalidated hourly.
```

Comment:
This is correct, but it also feels like premature optimization. You're about to do an API call, and the size of the map is limited, so the cost of iteration is probably negligible.
Feel free to ignore this comment if you disagree, as I said you're right and it's just a stylistic difference.